### PR TITLE
Take expected type into account when typing a sequence argument

### PR DESCRIPTION
**Pull request by smarter**

### CODE REVIEW

**Comment by smarter** [comment]
Hmmph, somehow this is causing scalatest to get stuck in an infinite loop, I'll investigate.

**Comment by smarter** [comment]
>  I'll investigate.

Fixed. I had written `if` somewhere instead of `else if` :).

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -726,20 +726,18 @@ class Typer extends Namer
...
-            arrayToRepeated(exprAdapted)
-          case _ =>
-            val starType = defn.SeqType.appliedTo(defn.AnyType)
-            val exprAdapted = adapt(tpdExpr, starType)
-            seqToRepeated(exprAdapted)
-        }
+        val expr1 = typedExpr(tree.expr, ptArg)
+        val fromCls = if expr1.tpe.derivesFrom(defn.ArrayClass) then defn.ArrayClass else defn.SeqClass
+        val tpt1 = TypeTree(expr1.tpe.widen.translateToRepeated(fromCls)).withSpan(tree.tpt.span)
+        assignType(cpy.Typed(tree)(expr1, tpt1), tpt1)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:740]: nice!

File: compiler/src/dotty/tools/dotc/core/TypeApplications.scala [hunk 2]
```diff
@@ -388,18 +388,24 @@ class TypeApplications(val self: Type) extends AnyVal {
...
-  def underlyingIfRepeated(isJava: Boolean)(implicit ctx: Context): Type =
-    if (self.isRepeatedParam) {
-      val seqClass = if (isJava) defn.ArrayClass else defn.SeqClass
-      // If `isJava` is set, then we want to turn `RepeatedParam[T]` into `Array[? <: T]`,
-      // since arrays aren't covariant until after erasure. See `tests/pos/i5140`.
-      translateParameterized(defn.RepeatedParamClass, seqClass, wildcardArg = isJava)
-    }
+  def translateFromRepeated(toArray: Boolean)(using Context): Type =
+    val seqClass = if (toArray) defn.ArrayClass else defn.SeqClass
+    if self eq WildcardType then
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/TypeApplications.scala:397]:
```suggestion
    if self.isinstanceOf[WildcardType] then
```
This works also if self is a bounded wildcard type.

**Pull request merged**